### DISCUSSION TITLE
NextJS 13 Vercel build failing. React-pdf

### DISCUSSION THREAD

**Discussion by ewaninfo** [vercel/community#3061] [Help] [answered]
### Summary

I'm using [react-pdf](https://www.npmjs.com/package/react-pdf). Everything works as expected locally.

Here is the error that is causing the Vercel build to fail:

`Error: /lib64/libc.so.6: version  'GLIBC_2.28' not found (required by /vercel/path0/pwa/node_modules/canvas/build/Release/libglib-2.0.so.0)`

Canvas is a dependency of pdf.js, which is a dependency of react-pdf.

Just wondering if anyone has experienced the same/similar issue? Any suggestions would be very welcome.

### Steps to Reproduce

Environment:
react: 18.2.0
next: 13.1.6
react-pdf: 7.1.2
node: 18.x

Create a NextJS app
Install react-pdf
Push the branch to Vercel

**Comment by smaeda-ks** [comment, marked as answer]
Hi, we're aware of this issue, and the only workaround as of today is to downgrade Node version to `16.x`. Could you try that?

We're actively discussing how we can support those GLIBC versions in Node `18.x`, and we will deliver updates when we have an update to share.

**Reply by ewaninfo** [reply]
Thanks for your reply. Downgrading the Node version brings new errors and concerns. Will keep an eye out for updates.

**Reply by Gnoll94** [reply]
Any updates on this? 16.x is reaching EOL in about a month, so would be nice to easily transition to 18.x whenever this gets fixed before builds stop working

**Reply by smaeda-ks** [reply]
We're going to extend the Node@16 deprecation date from the one we initially planned in this announcement:
https://vercel.com/changelog/node-js-14-and-16-are-being-deprecated

The exact date is still TBD, but we will likely postpone it for a few months. Will keep you updated.

**Reply by smaeda-ks** [reply]
We've updated the original plan and the new date is here:
https://vercel.com/changelog/node-js-14-and-16-are-being-deprecated

We will continue to support Node `16.x` until **February 6th 2024**

**Reply by nfts2me** [reply]
@smaeda-ks , can we expect this issue to be solved before the EOL?
This is preventing us from upgrading to next 14.

**Reply by jkingstoncsecond** [reply]
any updates on this @smaeda-ks ? using a 16.x.x version of node is causing other dependency issues in one of our projects, thanks

**Reply by raytawil-masterfulai** [reply]
Adding my voice to the thread, we're currently using 'canvas' and are unable to upgrade past node 16.x to node 18x. which is also starting to cause issues with the other deps in the app

**Reply by lancejpollard** [reply]
@smaeda-ks FWIW I was able to get node-canvas to build on Node 20 **but not Node 18** on Vercel. However, Node 20 has issues with `yum install x` (installing any yum package) in a custom install script, so there are other problems with v20.

Any update on getting it to work with Node 18?

**Reply by tisonkun** [reply]
@smaeda-ks is there a doc page telling which GLIBC versions are supported?

**Reply by j-wing** [reply]
Hi @smaeda-ks, wanted to check in on this. It sounds like the only options right now are to downgrade to a deprecated node version or upgrade to a beta one. Any news on plans to bring this fix to Node v18, since it's the only stable option right now?

**Reply by qianrum** [reply]
@smaeda-ks I also want to check in on this. Given that new builds for Node.js 16.x will no longer be available after June 14, 2014, I wonder if you have plans to address this issue. Really appreciate your help.

**Reply by 2pmflow** [reply]
Hi @smaeda-ks , wondering if there were any updates here, given node16 being disabled in less than two months. Appreciate your help!

**Comment by AidanLaycock** [comment]
We are running into the same issue (Not due to this specific package). Is there a timeline on when the image will be updated as EoL is about 6-7 weeks away now (https://vercel.com/changelog/node-js-14-and-16-are-being-deprecated) on Vercel.

As this is preventing us from upgrading some other services as well.

**Reply by smaeda-ks** [reply]
Right, totally valid concern, I agree. Extending the Node 16.x support is also on our table but it's hasn't decided yet. We will keep you updated in future updates. Thank you for your patience.

**Reply by AidanLaycock** [reply]
We have found an odd issue with this when trying to spin up the same codebase on a new vercel project (As we wanted to test that it wasn't an upgrade issue). And we did find that for one of our projects it did actually work with Node 18 on a new project. So I am not sure if this is also related to a pool of images, with some having issues rather than all of them.

Do let us know if there is anything we can do to help, as we want to see this fixed as soon as possible.

**Comment by weidezhang** [comment]
running into the same issue. also due to the canvas lib required by pdf-js.

**Reply by lancejpollard** [reply]
Same, using node-canvas via react-pdf.

**Comment by digredeth** [comment]
Im running into the same issue, the problem is that I cant downgrade node js version since I'm using langchain and it require node >= 18

**Reply by digredeth** [reply]
I'm running trough Render right now

**Comment by ethtea** [comment]
+1, also running into 

```
Error: /lib64/libc.so.6: version  'GLIBC_2.28' not found (required by /vercel/path0/pwa/node_modules/canvas/build/Release/libglib-2.0.so.0)
```

after upgrading to node 18, currently downgrading to 16

**Comment by wewlad-dev** [comment]
Same issue when building Zola 0.17.2 on vercel, I don't use Node but still have the problem.
`zola: /lib64/libm.so.6: version GLIBC_2.27 not found (required by zola)`

Folks at Zola said the image used by vercel is an old image https://github.com/getzola/zola/issues/1713#issuecomment-1097941400
> Vercel's builders run in an amazonlinux container, which uses outdated glibc. I tried to build zola 0.15.3 from source in that environment, and it works fine.

Update: I used Netlify instead

**Comment by jameslawson** [comment]
I also encountered a similar error when starting my first Vercel project, I have a monorepo and depend on [`@moonrepo/cli`](https://moonrepo.dev/).
My monorepo scripts/commands currently target Node v20.4.0, pnpm v8.6.7


```
/vercel/path0/node_modules/.bin/../@moonrepo/cli/moon: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /vercel/path0/node_modules/.bin/../@moonrepo/cli/moon)
--
09:13:26.509 | /vercel/path0/node_modules/.bin/../@moonrepo/cli/moon: /lib64/libc.so.6: version `GLIBC_2.29' not found (required by /vercel/path0/node_modules/.bin/../@moonrepo/cli/moon)
09:13:26.510 | /vercel/path0/node_modules/.bin/../@moonrepo/cli/moon: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by /vercel/path0/node_modules/.bin/../@moonrepo/cli/moon)
09:13:26.510 | /vercel/path0/node_modules/.bin/../@moonrepo/cli/moon: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by /vercel/path0/node_modules/.bin/../@moonrepo/cli/moon)
09:13:26.524 | Error: Command "npx moon run jx-frontend:build-production" exited with 1
```

I don't want to go through my monorepo and downgrade everything. I might try and deploy via a GitHub action as GitHub's machines support latest node for running my build commands. i.e. going to try this next: https://vercel.com/guides/how-can-i-use-github-actions-with-vercel#configuring-github-actions-for-vercel then probably give up on vercel unless anyone has any ideas

**Comment by pgzmnk** [comment]
Facing a similar issue when deploying a nextjs app to Vercel. The API routes that depend on the duckdb library error, while the other API routes work successfully. Downgraded node from 18.x to 16.x and the error persists.

```bash
Error: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /var/task/node_modules/duckdb/lib/binding/duckdb.node)
```

<img width="749" alt="image" src="https://github.com/vercel/community/assets/27398253/a320256d-57d4-4eb1-a792-42a8b4e631b0">

Thoughts appreciated.

**Reply by pgzmnk** [reply]
Disregard. This seems to be duckdb specific: https://github.com/duckdb/duckdb/issues/7088

**Comment by maderesponsively** [comment]
I'm receiving the same 'Error: /lib64/libc.so.6: version `GLIBC_2.28' not found' error. I'm using Next.js and Canvas. It's a bit worrying to see that v16 is going to be deprecated early next year. I don't want to have to move from Vercel hosting as it's great.

**Comment by wi-ski** [comment]
Deal breaker atm

**Comment by jbednar97** [comment]
Issue resolved with [this](https://github.com/ivansevillaa/use-next-blurhash/issues/4#issue-1095916956), but had similar issue with Node version. Using 16x fixed the problem, but it's really not great to see when the 16x version is about do be deprecated really soon

**Comment by abhinavtalentinc** [comment]
Hi @smaeda-ks ,

We have a use-case similar to @ewaninfo and some of the other members in this discussion. The workaround of using 16x has also worked for us for upgrading from 14x. We would appreciate if you could update the thread once a fix has been planned for this issue, so that we can plan upgrading our app to 18x accordingly.

**Comment by joost-berg** [comment]
Any update on this issue? I really want to use node 18 asap.

**Comment by dataconcept-it** [comment]
@smaeda-ks Any update on this issue? We are running into more and more issues because we cannot update our app to node 18x. The deprecation date is also not too far away anymore. Thanks in advance :)

**Comment by joost-berg** [comment]
Updating to nextjs 14 is also not possible with node 16x, so the problem is getting more and more urgent.

**Comment by mohaahadji** [comment]
I am getting the same issue! any update please?

**Reply by emirhangunec** [reply]
I'm waiting too, using fabric@5.3.0, canvas@2.8.0 first I get `ZLIB_1.2.9` error and fixed with adding `LD_LIBRARY_PATH=/vercel/path0/node_modules/canvas/build/Release:/var/task/node_modules/canvas/build/Release` env variable to vercel. 
Now I am getting `Error: /lib64/libc.so.6: version 'GLIBC_2.28' not found (required by /vercel/path0/node_modules/canvas/build/Release/libglib-2.0.so.0)` error and I can't find any solutions except downgrading to next 13 and node 16

**Reply by mohaahadji** [reply]
Yes that's exactly what's happened to me, still cannot figure out the issue!

**Reply by emirhangunec** [reply]
Also I can build with no errors on my M1 Macbook Pro runs on macOS sonoma 14

**Reply by Karyum** [reply]
@mohaahadji  I also had the same issue, until I noticed today Vercel has a new Node version currently in beta test (20.x).  
That fixed the issue for me, hope it helps

**Reply by jkingstoncsecond** [reply]
@Karyum do we need to change any settings in our vercel project or can we just enable it in the engines section of our packages.json?

**Reply by mohaahadji** [reply]
@Karyum just updated it and it works fine! Thanks for update.

**Comment by dataconcept-it** [comment]
@Karyum This works, thanks for the heads-up. It is finally resolved everyone! Very happy. 

Steps to fix:
1. Click on your project on Vercel
2. Click on tab Settings
3. Change the Node Version in section Node.js Version to **20.x**
4. Rebuild manually or commit new code to your project to deploy

I tested it with nextjs 14x and all packages, **including react-pdf**, updated. And it works.

Very happy, thanks to everyone for keeping this issue alive and for resolving it.

**Reply by dataconcept-it** [reply]
@jkingstonc this is how to fix it.

**Reply by jkingstoncsecond** [reply]
thanks @dataconcept-it, i have just tried this however and vercel is complaining about 20.x.x not being supported. I have changed the version to 20.x.x in my project settings, re-deployed and am getting the following error below. I have then also specified the version in my `package.json` to see if that fixed it `"engines": {"node": ">20.0.0"},` and same error. is anyone else having this issue?

`Error: Found invalid Node.js Version: ">20.0.0". Please set "engines": { "node": "18.x" } in your `package.json` file to use Node.js 18.`

EDIT: Fixed, for anyone who has previously tried to fix this by using the `VERCEL_CLI_VERSION` environment variable suggested in other threads, I have removed this and now can use node 20.x.x

**Reply by dataconcept-it** [reply]
Can you remove the entire entry `"engines": {"node": ">20.0.0"},` from your package.json file? I do not have this and it works without issues. I would try the following steps:
1. On your local machine, make sure node v20x is installed
2. Remove `"engines": {"node": ">20.0.0"},` from package.json
3. Remove node_modules (`rm -rf node_modules`)
4. Reinstall packages (yarn/npm)
5. Run `next build` and see if that works. 

If all works well, redeploy to vercel and see what happens

**Reply by dataconcept-it** [reply]
Ah, you fixed it already!

**Reply by jkingstoncsecond** [reply]
yep all working now appreciate the help!

**Comment by abhinavtalentinc** [comment]
Hello @smaeda-ks ,

We are seeing that a few users are not facing this issue on using Node version 20 in Vercel. Can you please confirm if the fix also going to be applicable for version 18, so that we can plan the upgrade accordingly?
